source/fuzz/transformation_propagate_instruction_up.cpp

```diff
@@ -0,0 +1,393 @@
+    case SpvOpIsNormal:
+    case SpvOpSignBitSet:
+    case SpvOpLessOrGreater:
+    case SpvOpOrdered:
+    case SpvOpUnordered:
+    case SpvOpLogicalEqual:
+    case SpvOpLogicalNotEqual:
+    case SpvOpLogicalOr:
+    case SpvOpLogicalAnd:
+    case SpvOpLogicalNot:
+    case SpvOpSelect:
+    case SpvOpIEqual:
+    case SpvOpINotEqual:
+    case SpvOpUGreaterThan:
+    case SpvOpSGreaterThan:
+    case SpvOpUGreaterThanEqual:
+    case SpvOpSGreaterThanEqual:
+    case SpvOpULessThan:
+    case SpvOpSLessThan:
+    case SpvOpULessThanEqual:
+    case SpvOpSLessThanEqual:
+    case SpvOpFOrdEqual:
+    case SpvOpFUnordEqual:
+    case SpvOpFOrdNotEqual:
+    case SpvOpFUnordNotEqual:
+    case SpvOpFOrdLessThan:
+    case SpvOpFUnordLessThan:
+    case SpvOpFOrdGreaterThan:
+    case SpvOpFUnordGreaterThan:
+    case SpvOpFOrdLessThanEqual:
+    case SpvOpFUnordLessThanEqual:
+    case SpvOpFOrdGreaterThanEqual:
+    case SpvOpFUnordGreaterThanEqual:
+    case SpvOpShiftRightLogical:
+    case SpvOpShiftRightArithmetic:
+    case SpvOpShiftLeftLogical:
+    case SpvOpBitwiseOr:
+    case SpvOpBitwiseXor:
+    case SpvOpBitwiseAnd:
+    case SpvOpNot:
+    case SpvOpBitFieldInsert:
+    case SpvOpBitFieldSExtract:
+    case SpvOpBitFieldUExtract:
+    case SpvOpBitReverse:
+    case SpvOpBitCount:
+    case SpvOpCopyLogical:
+    case SpvOpPtrEqual:
+    case SpvOpPtrNotEqual:
+      return true;
+    default:
+      return false;
+  }
+}
+
+opt::Instruction*
+TransformationPropagateInstructionUp::GetInstructionToPropagate(
+    opt::IRContext* ir_context, uint32_t block_id) {
+  auto* block = ir_context->cfg()->block(block_id);
+  assert(block && "|block_id| is invalid");
+
+  for (auto& inst : *block) {
+    if (inst.opcode() == SpvOpPhi || !IsOpcodeSupported(inst.opcode())) {
+      continue;
+    }
+
+    auto valid = true;
+    for (uint32_t i = 0; i < inst.NumInOperands(); ++i) {
+      const auto& operand = inst.GetInOperand(i);
+      if (operand.type != SPV_OPERAND_TYPE_ID) {
+        continue;
+      }
+
+      auto* dependency =
+          ir_context->get_def_use_mgr()->GetDef(operand.words[0]);
+      assert(dependency && "Operand has invalid id");
+
+      if (ir_context->get_instr_block(dependency) == block &&
+          dependency->opcode() != SpvOpPhi) {
+        valid = false;
+        break;
+      }
+    }
+
+    if (valid) {
+      return &inst;
+    }
+  }
+
+  return nullptr;
+}
+
+bool TransformationPropagateInstructionUp::IsApplicableToTheBlock(
+    opt::IRContext* ir_context, uint32_t block_id) {
+  // Check that |block_id| is valid.
+  const auto* block = ir_context->cfg()->block(block_id);
+  if (!block) {
+    return false;
+  }
+
+  // Check that |block| is reachable in the CFG.
+  if (ir_context->cfg()->preds(block_id).empty()) {
+    return false;
+  }
+
+  // |block_id| can't be the id of the first block in the function.
+  if (block_id == block->GetParent()->begin()->id()) {
+    return false;
+  }
```

Comment:
Isn't this captured by the previous check?  I'm assuming the first block doesn't have any predecessors.  (Unless the cfg analysis includes some sort of placeholder predecessor.)

Reply:
I've removed this check in 51c3f43. I'm not sure if the first block can't have any predecessors (e.g. if it's a loop header).

Reply:
Yep, from the section 2.16 of SPIR-V spec:
```
The first block in a function definition is the entry point of that function and cannot be the target of any branch. (Note this means it will have no OpPhi instructions.)
```